compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -66,35 +70,18 @@ fn anon_const_type_of<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Ty<'tcx> {
                 .expect("const parameter types cannot be generic");
         }
 
-        Node::TypeBinding(binding @ &TypeBinding { hir_id: binding_id, .. })
-            if let Node::TraitRef(trait_ref) = tcx.hir_node(tcx.hir().parent_id(binding_id)) =>
-        {
-            let Some(trait_def_id) = trait_ref.trait_def_id() else {
-                return Ty::new_error_with_message(
-                    tcx,
-                    tcx.def_span(def_id),
-                    "Could not find trait",
-                );
-            };
-            let assoc_items = tcx.associated_items(trait_def_id);
-            let assoc_item = assoc_items.find_by_name_and_kind(
-                tcx,
-                binding.ident,
-                ty::AssocKind::Const,
-                def_id.to_def_id(),
-            );
-            return if let Some(assoc_item) = assoc_item {
-                tcx.type_of(assoc_item.def_id)
-                    .no_bound_vars()
-                    .expect("const parameter types cannot be generic")
-            } else {
-                // FIXME(associated_const_equality): add a useful error message here.
-                Ty::new_error_with_message(
-                    tcx,
-                    tcx.def_span(def_id),
-                    "Could not find associated const on trait",
-                )
+        Node::TypeBinding(&TypeBinding { hir_id, ident, .. }) => {
+            let ty = tcx.type_of_assoc_const_binding(hir_id).skip_binder().skip_binder();
+
+            // We can't possibly catch this in the resolver, therefore we need to handle it here.
+            // FIXME(const_generics): Support generic const generics.
+            if ty.has_param() || ty.has_escaping_bound_vars() {
```

Comment:
Feels a bit weird to do perform all those checks here in `type_of`. I can move them as well as the query fallback to `astconv/bounds.rs` and the of course change the return type of the query from `ty::EarlyBinder<ty::Binder<'tcx, Ty<'tcx>>>` to just `Ty<'tcx>`. There shouldn't be any observable change in behavior.